Added a visitor version of unpack API,

redboltz opened the pull request:
The current unpacking APIs are constructed on the visitor mechanism.
(Fixed #418)

redboltz commented:
There is no significant difference in performance.
![image](https://cloud.githubusercontent.com/assets/275959/14945807/581c3fc4-1053-11e6-94bc-2782b3015171.png)

jpetso commented on test/unpack_visitor.cpp:
```diff
@@ -0,0 +1,135 @@
+
+#if MSGPACK_DEFAULT_API_VERSION >= 2
+
+struct json_visitor : msgpack::v2::null_visitor {
```
Since this test is also listed on the wiki documentation page, I'll point out two potential issues:
* `end_map_value()` doesn't append a comma to separate map items like `end_array_item()` does.
* `visit_str()` doesn't process strings as Unicode, therefore non-ASCII codepoints will result in invalid JSON.

Neither of these is required by the test case so that's okay for this particular piece of code, however you might want to point it out in the documentation.

redboltz replied:
Thank you for the comments.

> end_map_value() doesn't append a comma to separate map items like end_array_item() does.

I will fix it.

> visit_str() doesn't process strings as Unicode, therefore non-ASCII codepoints will result in invalid JSON (which encodes them as e.g. "\u00f8").

True. The point of the test and the document is not accurate JSON, I just demonstrate a visitor usage. However, I should use a precise term. I will fix the test and the document. I will use the term 'json like structure ( string is not espaced)'.

jpetso commented:
One more thing to consider: When I was having a look at https://github.com/msgpack/msgpack-c/wiki/v2_0_cpp_unpack_visit, I realized that this code isn't actually unpacking the msgpack buffer, at least not in the traditional sense of unpack as "parse buffer into msgpack::object".

Since the function name (`unpack_visit()`) is already not the same as the traditional `unpack()`, I wonder whether it makes sense to give it a name that's less tied to the traditional meaning of "unpack". Maybe `parse()`? I'm interested to hear about your thoughts on this idea :)

redboltz commented:
Good point. `unpack_visit()` doesn't unpack anything. I think that `parse()` is good too. Another candidate is `traverse()`. In this context, `parse()` is better. I will rename to `parse()`.

redboltz commented:
I've fixed all problems that you pointed out. In addition, I found end_map_value() calling point was wrong, and fixed it. https://github.com/redboltz/msgpack-c/blob/add_unpack_visitor_api/include/msgpack/v2/unpack.hpp#L434

redboltz commented:
@jpetso , could you review my fix? If my fix would be ok, I will merge it.

redboltz commented:
I am going to merge the PR  this weekend if no comments are post.

jpetso commented:
Hi, sorry, and well done! I had a pretty tight few weeks and didn't manage to find the time for the review. Since the first version of the patch was already very good, I'm assuming it's quite alright :)

Anyway, I'm glad it's in, and thank you for your hard work!

redboltz commented:
@jpetso , thank you!
